packages/zoe/src/contractFacet.js

```diff
@@ -351,45 +241,43 @@ export function buildRootObject(_vatPowers) {
       getBrandForIssuer: issuer => issuerTable.brandFromIssuer(issuer),
       getAmountMath: getAmountMathForBrand,
       getVatAdmin: instanceRecord.adminNode,
-    });
+    };
+    return harden(contractFacet);
   };
 
   /**
-   * @typedef {import('@agoric/zoe').CompleteObj} CompleteObj
-   *
-   * @typedef {Object} ZcfForZoe
-   * The facet ZCF presents to Zoe.
-   *
-   * @property {(OfferHandle, Proposal, Allocation) => CompleteObj} addOffer
-   * Add a single offer to this contract instance.
-   *
-   * @return CompleteObj
+   * @returns {ZcfForZoe}
    */
-  const makeZcfForZoe = (instanceHandle, zoeForZcf) =>
-    harden({
+  const makeZcfForZoe = (instanceHandle, zoeForZcf) => {
+    /** @type {ZcfForZoe} */
+    const zcfForZoe = {
       addOffer: (offerHandle, proposal, allocation) => {
         const ignoringUpdater = harden({
           updateState: () => {},
           resolve: () => {},
         });
+
+        const cleanedProposal = cleanProposal(getAmountMathForBrand, proposal);
         const offerRecord = {
           instanceHandle,
-          proposal,
+          proposal: cleanedProposal,
           currentAllocation: allocation,
           notifier: undefined,
           updater: ignoringUpdater,
         };
 
-        const { exit } = proposal;
+        const { exit } = cleanedProposal;
         const [exitKind] = Object.getOwnPropertyNames(exit);
 
+        /** @type {CompleteObj | undefined} */
         let completeObj;
         const completeOffer = () => {
           return completeOffers(harden([offerHandle]), zoeForZcf);
         };
 
         if (exitKind === 'afterDeadline') {
           // Automatically complete offer after deadline.
+          assert(exit.afterDeadline);
```

Comment:
I did this entirely to appease the type checker so that I could dereference the subproperties.  I'd prefer not to go further, but yes, we could do more sanity checking.